### PR TITLE
add applications and services namespace create

### PR DESCRIPTION
**Pull request by mmulholla**
The code to create namespaces runs successfully but so far has not been tested to check it all wires up correctly. 

Sample invocations:

```out/venv3/bin/primazactl worker create applications-namespace -c kind-primazactl-worker-test -m kind-primazactl-main-test -d test-environment ```

```out/venv3/bin/primazactl worker create services-namespace -c kind-primazactl-worker-test -m kind-primazactl-main-test -d test-environment ```

main logic is in ```primazactl/scripts/src/primazactl/primazaworker/workernamespace.py create() function```  with comments added to match steps defined in the jira issue: https://issues.redhat.com/browse/APPSVC-1283

still to do: 
- add checks as defined in the jira issue (will use separate PR)
- README updates (will use separate PR)

### CODE REVIEW

File: scripts/src/primazactl/cmd/main/uninstall.py [hunk 1]
```diff
@@ -14,9 +16,14 @@ def add_uninstall(parser: argparse.ArgumentParser, parents=[]):
...
+        PrimazaMain(
+            args.cluster_name,
+            args.kubeconfig,
+            args.config,
+            args.version).uninstall_primaza()
+    except Exception as e:
+        print(traceback.format_exc())
+        print(f"\nAn exception occurred executing main install: {e}",
+              file=sys.stderr)
+    print("Primaza main successfully uninstalled")
```
Comment (filariow) [inline scripts/src/primazactl/cmd/main/uninstall.py]:
```suggestion
            args.version).uninstall_primaza()
            print("Primaza main successfully uninstalled")
    except Exception as e:
        print(traceback.format_exc())
        print(f"\nAn exception occurred executing main install: {e}",
              file=sys.stderr)
```

File: scripts/src/primazactl/cmd/worker/create/namespace.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import sys
+from primazactl.types import kubernetes_name
+from primazactl.primazaworker.workernamespace import WorkerNamespace
+from .constants import SERVICES, APPLICATIONS
+
+
+def add_create_applications_namespace(parser: argparse.ArgumentParser,
+                                      parents=[]):
+    applications_namespace_parser = parser.add_parser(
+        "applications-namespace",
```
Comment (filariow) [inline scripts/src/primazactl/cmd/worker/create/namespace.py]: I'd propose to use the singular form here. This way the final command would be `primazactl main create application-namespace`.

```suggestion
        "application-namespace",
```

File: scripts/src/primazactl/cmd/worker/create/namespace.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        parents=parents)
+    applications_namespace_parser.set_defaults(
+        func=create_applications_namespace)
+    add_args_namespace(applications_namespace_parser)
+
+
+def add_create_services_namespace(parser: argparse.ArgumentParser,
+                                  parents=[]):
+    services_namespace_parser = parser.add_parser(
+        "services-namespace",
```
Comment (filariow) [inline scripts/src/primazactl/cmd/worker/create/namespace.py]: as above

```suggestion
        "service-namespace",
```

File: scripts/src/primazactl/cmd/worker/create/namespace.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        help="name of cluster, as it appears in kubeconfig, \
+                on which Primaza is installed. Default: \
+                current kubeconfig context",
+        type=str,
+        default=None)
+
+
+def create_applications_namespace(args):
+    try:
+        namespace = WorkerNamespace(APPLICATIONS,
```
Comment (filariow) [inline scripts/src/primazactl/cmd/worker/create/namespace.py]: eventually the namespace name should be a parameter

File: scripts/src/primazactl/kube/namespace.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        logger.log_entry(f"namespace: {self.name}")
+
+        try:
+            return self.corev1.read_namespace(name=self.name)
+        except ApiException as e:
+            if e.reason != "Not Found":
+                logger.log_error("Exception when calling "
+                                 "read_namespace: %s\n" % e)
+                raise e
+        return ""
```
Comment (filariow) [inline scripts/src/primazactl/kube/namespace.py]: It seems to me that return type should be `V1Namespace | None` (cfr. https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/CoreV1Api.md#read_namespace). This should also apply to other classes in `kube` module.

```suggestion
    def read(self) -> client.V1Namespace | None:
        logger.log_entry(f"namespace: {self.name}")

        try:
            return self.corev1.read_namespace(name=self.name)
        except ApiException as e:
            if e.reason != "Not Found":
                logger.log_error("Exception when calling "
                                 "read_namespace: %s\n" % e)
                raise e
        return None
```

File: scripts/src/primazactl/primaza/primaza.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+        docker_data = yaml.safe_load(out)
+        try:
+            networks = docker_data[0]["NetworkSettings"]["Networks"]
+            ipaddr = networks["kind"]["IPAddress"]
+            logger.log_info(f"new cluster url: https://{ipaddr}:6443")
+            return f"https://{ipaddr}:6443"
+        except KeyError:
+            logger.log_info("new cluster url not found")
+            return ""
```
Comment (filariow) [inline scripts/src/primazactl/primaza/primaza.py]: eventually we should remove this kind-related logic from `primazactl`. I'd propose to add a new flag the user can set to override the serverUrl in generated kubeconfig.

However, that can be done in another PR. As of now I'd say we can add a `TODO` comment here to not forget to remove this code.

File: scripts/src/primazactl/primaza/primaza.py
Comment (filariow) [inline scripts/src/primazactl/primaza/primaza.py:1]: Had a similar naming and a problems on imports very hard to debug. When you have the folder and the file with the same name python has some strange behavior. I suggest to rename or the folder or the file.

File: scripts/src/primazactl/primaza/primaza.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
+import yaml
+from typing import Dict
+from primazactl.utils import logger, command
+from primazactl.identity import kubeidentity
+from primazactl.kube.secret import Secret
+from primazactl.utils import kubeconfig
+from primazactl.utils.kubeconfigwrapper import KubeConfigWrapper
+
+
+class Primaza(object):
```
Comment (filariow) [inline scripts/src/primazactl/primaza/primaza.py]: may `Cluster` or `PrimazaCluster` be a more appropriate name?

```suggestion
class PrimazaCluster(object):
```

File: scripts/src/primazactl/cmd/worker/create/namespace.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                                    args.cluster_environment,
+                                    args.cluster_name,
+                                    args.main_clustername)
+        namespace.create()
+        print("Applications namespace was successfully created")
+    except Exception:
+        print(traceback.format_exc())
+        print("\nAn exception creating an applications namespace",
+              file=sys.stderr)
```
Comment (filariow) [inline scripts/src/primazactl/cmd/worker/create/namespace.py]: we should also create a service account on Primaza Main, and create in the worker namespace a secret with the kubeconfig. That also applies to Service Namespaces. Something like the following:

```suggestion
def create_applications_namespace(args):
    try:
        main = PrimazaMain(.....)
        sa = main.create_service_account(args.cluster_environment)
        kcfg = sa.get_kubeconfig()
              
        namespace = WorkerNamespace(APPLICATIONS,
                                    args.cluster_environment,
                                    args.cluster_name,
                                    args.main_clustername)
        namespace.create()
        
        worker = PrimazaWorker(....)
        worker.create_primaza_secret(APPLICATIONS, kcfg)
        
        print("Applications namespace was successfully created")
    except Exception:
        print(traceback.format_exc())
        print("\nAn exception creating an applications namespace",
              file=sys.stderr)
```

**Review by filariow — CHANGES_REQUESTED** [review]
some feedbacks

File: scripts/src/primazactl/cmd/worker/create/namespace.py (hunk 2 above)
Comment (mmulholla) [reply scripts/src/primazactl/cmd/worker/create/namespace.py]: Fixed

File: scripts/src/primazactl/cmd/main/uninstall.py (hunk 1 above)
Comment (mmulholla) [reply scripts/src/primazactl/cmd/main/uninstall.py]: Fixed.

File: scripts/src/primazactl/cmd/worker/create/namespace.py (hunk 3 above)
Comment (mmulholla) [reply scripts/src/primazactl/cmd/worker/create/namespace.py]: Fixed

File: scripts/src/primazactl/cmd/worker/create/namespace.py (hunk 4 above)
Comment (mmulholla) [reply scripts/src/primazactl/cmd/worker/create/namespace.py]: updated the code to make it easier in the future.

File: scripts/src/primazactl/kube/namespace.py (hunk 5 above)
Comment (mmulholla) [reply scripts/src/primazactl/kube/namespace.py]: Fixed

File: scripts/src/primazactl/primaza/primaza.py (hunk 6 above)
Comment (mmulholla) [reply scripts/src/primazactl/primaza/primaza.py]: Agreed.

File: scripts/src/primazactl/primaza/primaza.py
Comment (mmulholla) [reply scripts/src/primazactl/primaza/primaza.py:1]: renamed all files with same name as containing directory.

File: scripts/src/primazactl/primaza/primaza.py (hunk 7 above)
Comment (mmulholla) [reply scripts/src/primazactl/primaza/primaza.py]: Sure, changed to that.

File: scripts/src/primazactl/cmd/worker/create/namespace.py (hunk 8 above)
Comment (mmulholla) [reply scripts/src/primazactl/cmd/worker/create/namespace.py]: Fixed.

**Review by dperaza4dustbit — DISMISSED** [review]
lgtm

File: scripts/src/primazatest/config/kind.yaml.bak
Comment (filariow) [inline scripts/src/primazatest/config/kind.yaml.bak:1]: is this file needed?

File: scripts/src/primazatest/config/kind.yamlbak
Comment (filariow) [inline scripts/src/primazatest/config/kind.yamlbak:1]: is this file needed?

File: scripts/src/primazactl/primazaworker/workernamespace.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        # Get kubeconfig with secret from service accounf
+        kc = self.main.get_kubeconfig(sa_name, self.cluster_name)
+
+        # - in the created namespace, create the Secret
+        #     'primaza-auth-$CLUSTER_ENVIRONMENT' the Worker key
+        #     and the kubeconfig for authenticating with the Primaza cluster.
+        if self.type == APPLICATION:
+            secret_name = "primaza-kubeconfig"
+        else:
+            secret_name = "primaza-kubeconfig"
```
Comment (filariow) [inline scripts/src/primazactl/primazaworker/workernamespace.py]: that seems redundant

**Review by filariow — CHANGES_REQUESTED** [review]
looks mostly good to me, a couple of nits

File: scripts/src/primazatest/config/kind.yaml.bak
Comment (mmulholla) [reply scripts/src/primazatest/config/kind.yaml.bak:1]: oops, removed it.

File: scripts/src/primazatest/config/kind.yamlbak
Comment (mmulholla) [reply scripts/src/primazatest/config/kind.yamlbak:1]: oops, removed it.

File: scripts/src/primazactl/primazaworker/workernamespace.py (hunk 9 above)
Comment (mmulholla) [reply scripts/src/primazactl/primazaworker/workernamespace.py]: Yes, moved the name to an object variable so the if..else is not needed at all.

**Review by filariow — APPROVED** [review]
lgtm
